pkg/yurthub/autonomymanager/autonomy_manager.go

```diff
@@ -0,0 +1,227 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autonomymanager
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	appsv1beta1 "github.com/openyurtio/openyurt/pkg/apis/apps/v1beta1"
+	"github.com/openyurtio/openyurt/pkg/projectinfo"
+	"github.com/openyurtio/openyurt/pkg/yurthub/cachemanager"
+	hubrest "github.com/openyurtio/openyurt/pkg/yurthub/kubernetes/rest"
+	"github.com/openyurtio/openyurt/pkg/yurthub/storage"
+)
+
+type AutonomyManager struct {
+	sync.RWMutex
+	cacheMgr      cachemanager.CacheManager
+	nodeName      string
+	count         int
+	queue         workqueue.RateLimitingInterface
+	errorKeys     map[storage.KeyBuildInfo]error
+	restConfigMgr *hubrest.RestConfigManager
+}
+
+func NewAutonomyManager(restConfigMgr *hubrest.RestConfigManager, cacheMgr cachemanager.CacheManager, nodeName string) (*AutonomyManager, error) {
+	am := &AutonomyManager{
+		nodeName:      nodeName,
+		count:         0,
+		cacheMgr:      cacheMgr,
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ""),
+		restConfigMgr: restConfigMgr,
+	}
+	return am, nil
+}
+
+// Run the autonomy manager
+// 1. Receive delta key
+// 2. Dispatch keys to recover routine
+// 3. Recover routine get object for api server
+// 4. Write it to cache manager and refresh error key
+// utilize queue and wait
+func (am *AutonomyManager) Run(ctx context.Context) {
+	// 1.Consistency Validate every 20 seconds
+	go wait.JitterUntil(func() {
+		if am.restConfigMgr == nil {
+			return
+		}
+		config := am.restConfigMgr.GetRestConfig(true)
+		if config == nil {
+			return
+		}
+		client, err := kubernetes.NewForConfig(config)
+		if err != nil {
+			return
+		}
+		am.ConsistencyValidate(client)
+	}, 20*time.Second, 2, true, ctx.Done())
+
+	// 2. Sync
+	go wait.JitterUntil(func() {
+		if am.restConfigMgr == nil {
+			return
+		}
+		config := am.restConfigMgr.GetRestConfig(true)
+		if config == nil {
+			return
+		}
+		dynamicClient, err := dynamic.NewForConfig(config)
+		if err != nil {
+			return
+		}
+		am.Sync(dynamicClient)
+	}, 20*time.Second, 3, true, ctx.Done())
+
+	// 3. Update errorKeys
+	observer := am.cacheMgr.ObserveErrorKeys()
+	go am.syncErrorKeys(ctx, observer)
+}
+
+// ConsistencyValidate check the data consistency between cache manager and api server
+func (am *AutonomyManager) ConsistencyValidate(client kubernetes.Interface) {
+	klog.Info("start to check the data consistency between cache manager and api server")
+	node, err := client.CoreV1().Nodes().Get(context.TODO(), am.nodeName, metav1.GetOptions{})
```

Comment:
It is a very high cost for Yurthub to get node data for every 20s, because cloud-edge traffic will rise a lot.